File: 22. Generate Parentheses.java
public class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> output = new ArrayList<>();
        make("", n, 0, 0, output);
        return output;
    }
    
    public void make(String sb, int n, int l, int r, List<String> output) {
        if (r == n) {
            output.add(sb);
            return;
        } else {
            if (r < l) {
                make(sb + ")", n, l, r + 1, output);
            }
            if (l < n) {
                make(sb + "(", n, l + 1, r, output);
            }
        }
    }
}

File: 22. Generate Parentheses.py
class Solution:
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        output = []
        
        def generate(inpt, left, right, n):
            if len(inpt) == 2 * n:
                output.append(inpt)
                return
            
            if left < right and left < n:
                generate(inpt + ')', left + 1, right, n)
            
            if right < n:
                generate(inpt + '(', left, right + 1, n)
        
        generate("", 0, 0, n)
        return output
